From 7ca559a03e8a843463d49a501e178d0623b541fc Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Thu, 4 May 2023 17:54:58 +0200
Subject: [PATCH 0392/1695] drivers: stm32_rng: allow RNG configurations to be
 defined in device tree

Having entropy source configuration explicitly defined in the RNG driver
is not easy to maintain and is not mandatory. Leave the ability to
the developer to specify an optional RNG entropy source configuration
in the device tree.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 0e43dbe4aa02ff4d17db5d7279d83ee0c6199f2f drivers: stm32_rng: allow RNG configurations to be defined in device tree]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: If1435acc796f211ae0d7cca616ca918cd1b95569
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/304380
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Etienne CARRIERE <etienne.carriere@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/stm32_rng.c | 72 +++++++++++++++++-----------------------
 1 file changed, 31 insertions(+), 41 deletions(-)

diff --git a/core/drivers/stm32_rng.c b/core/drivers/stm32_rng.c
index c88d41151..c53c2359d 100644
--- a/core/drivers/stm32_rng.c
+++ b/core/drivers/stm32_rng.c
@@ -32,9 +32,15 @@
 #define RNG_CR_RNGEN		BIT(2)
 #define RNG_CR_IE		BIT(3)
 #define RNG_CR_CED		BIT(5)
+#define RNG_CR_CONFIG1		GENMASK_32(11, 8)
+#define RNG_CR_NISTC		BIT(12)
+#define RNG_CR_CONFIG2		GENMASK_32(15, 13)
 #define RNG_CR_CLKDIV		GENMASK_32(19, 16)
 #define RNG_CR_CLKDIV_SHIFT	U(16)
+#define RNG_CR_CONFIG3		GENMASK_32(25, 20)
 #define RNG_CR_CONDRST		BIT(30)
+#define RNG_CR_ENTROPY_SRC_MASK	(RNG_CR_CONFIG1 | RNG_CR_NISTC | \
+				 RNG_CR_CONFIG2 | RNG_CR_CONFIG3)
 
 #define RNG_SR_DRDY		BIT(0)
 #define RNG_SR_CECS		BIT(1)
@@ -42,6 +48,8 @@
 #define RNG_SR_CEIS		BIT(5)
 #define RNG_SR_SEIS		BIT(6)
 
+#define RNG_NSCR_MASK		GENMASK_32(17, 0)
+
 #if TRACE_LEVEL > TRACE_DEBUG
 #define RNG_READY_TIMEOUT_US	U(100000)
 #else
@@ -50,22 +58,15 @@
 #define RNG_RESET_TIMEOUT_US	U(1000)
 
 #define RNG_FIFO_BYTE_DEPTH	U(16)
+#define RNG_CONF_LEN		U(3)
 
-#define RNG_NIST_CONFIG_A	U(0x0F00F00)
-#define RNG_HEALTH_CONFIG_A	U(0xAB39)
-#define RNG_NOISE_CTRL_CONFIG_A	U(0x0492)
-#define RNG_NIST_CONFIG_B	U(0x1801000)
-#define RNG_HEALTH_CONFIG_B	U(0xAAC7)
-#define RNG_NOISE_CTRL_CONFIG_B	U(0x3FFFF)
-#define RNG_NIST_CONFIG_C	U(0x0F00D00)
-#define RNG_HEALTH_CONFIG_C	U(0xAAC7)
-#define RNG_NOISE_CTRL_CONFIG_C	U(0x3FFFF)
 #define RNG_NIST_CONFIG_MASK	GENMASK_32(25, 8)
 
 #define RNG_MAX_NOISE_CLK_FREQ	U(3000000)
 
 struct stm32_rng_driver_data {
 	bool has_cond_reset;
+	bool entropy_src_config;
 };
 
 struct stm32_rng_instance {
@@ -481,7 +482,7 @@ static TEE_Result stm32_rng_parse_fdt(const void *fdt, int node)
 	TEE_Result res = TEE_ERROR_GENERIC;
 	const fdt32_t *cuint = NULL;
 	struct dt_node_info dt_rng = { };
-	unsigned int config_id = 0;
+	int len = 0;
 
 	fdt_fill_device_info(fdt, &dt_rng, node);
 	if (dt_rng.reg == DT_INFO_INVALID_REG)
@@ -507,40 +508,29 @@ static TEE_Result stm32_rng_parse_fdt(const void *fdt, int node)
 	stm32_rng->release_post_boot = IS_ENABLED(CFG_WITH_SOFTWARE_PRNG) &&
 				       !IS_ENABLED(CFG_PM);
 
-	if (!stm32_rng->ddata->has_cond_reset)
-		goto end;
+	if (!stm32_rng->ddata->entropy_src_config)
+		return TEE_SUCCESS;
 
-	cuint = fdt_getprop(fdt, node, "st,rng-config", NULL);
+	cuint = fdt_getprop(fdt, node, "st,rng-entropy-source-config", &len);
 	if (!cuint) {
-		IMSG("No RNG configuration specified, defaulting to NIST B");
-		stm32_rng->rng_config = RNG_NIST_CONFIG_B;
-		stm32_rng->health_test_conf = RNG_HEALTH_CONFIG_B;
-		stm32_rng->noise_ctrl_conf = RNG_NOISE_CTRL_CONFIG_B;
-		goto end;
-	}
+		if (len != -FDT_ERR_NOTFOUND)
+			panic();
 
-	config_id = (unsigned int)fdt32_to_cpu(*cuint);
-	switch (config_id) {
-	case 0:
-		stm32_rng->rng_config = RNG_NIST_CONFIG_A;
-		stm32_rng->health_test_conf = RNG_HEALTH_CONFIG_A;
-		stm32_rng->noise_ctrl_conf = RNG_NOISE_CTRL_CONFIG_A;
-		break;
-	case 1:
-		stm32_rng->rng_config = RNG_NIST_CONFIG_B;
-		stm32_rng->health_test_conf = RNG_HEALTH_CONFIG_B;
-		stm32_rng->noise_ctrl_conf = RNG_NOISE_CTRL_CONFIG_B;
-		break;
-	case 2:
-		stm32_rng->rng_config = RNG_NIST_CONFIG_C;
-		stm32_rng->health_test_conf = RNG_HEALTH_CONFIG_C;
-		stm32_rng->noise_ctrl_conf = RNG_NOISE_CTRL_CONFIG_C;
-		break;
-	default:
-		panic("RNG config not supported");
+		IMSG("No RNG configuration specified, preserving default one");
+
+		return TEE_SUCCESS;
+	} else if (len / sizeof(uint32_t) != RNG_CONF_LEN) {
+		panic("Incorrect number of arguments for RNG configuration");
 	}
 
-end:
+	stm32_rng->rng_config = fdt32_to_cpu(cuint[0]);
+	if (stm32_rng->rng_config & ~RNG_CR_ENTROPY_SRC_MASK)
+		panic("Incorrect entropy source configuration");
+	stm32_rng->health_test_conf = fdt32_to_cpu(cuint[1]);
+	stm32_rng->noise_ctrl_conf = fdt32_to_cpu(cuint[2]);
+	if (stm32_rng->noise_ctrl_conf & ~RNG_NSCR_MASK)
+		panic("Incorrect noise source control configuration");
+
 	return TEE_SUCCESS;
 }
 
@@ -604,11 +594,11 @@ err:
 }
 
 static const struct stm32_rng_driver_data mp13_data[] = {
-	{ .has_cond_reset = true },
+	{ .has_cond_reset = true, .entropy_src_config = true },
 };
 
 static const struct stm32_rng_driver_data mp15_data[] = {
-	{ .has_cond_reset = false },
+	{ .has_cond_reset = false, .entropy_src_config = false },
 };
 DECLARE_KEEP_PAGER(mp15_data);
 
-- 
2.47.2

